ambry-commons/src/main/java/com.github.ambry.commons/BlobId.java

```diff
@@ -166,30 +167,12 @@ private BlobId(short version, BlobIdType type, byte datacenterId, short accountI
    * @throws IOException
    */
   private BlobId(DataInputStream stream, ClusterMap clusterMap, boolean ensureFullyRead) throws IOException {
-    version = stream.readShort();
-    switch (version) {
-      case BLOB_ID_V1:
-        type = BlobIdType.NATIVE;
-        datacenterId = UNKNOWN_DATACENTER_ID;
-        accountId = UNKNOWN_ACCOUNT_ID;
-        containerId = UNKNOWN_CONTAINER_ID;
-        break;
-      case BLOB_ID_V2:
-        stream.readByte();
-        type = BlobIdType.NATIVE;
-        datacenterId = stream.readByte();
-        accountId = stream.readShort();
-        containerId = stream.readShort();
-        break;
-      case BLOB_ID_V3:
-        type = BlobIdType.values()[stream.readByte() & 0x3];
-        datacenterId = stream.readByte();
-        accountId = stream.readShort();
-        containerId = stream.readShort();
-        break;
-      default:
-        throw new IllegalArgumentException("blobId version " + version + " not supported.");
-    }
+    BlobIdPreamble preamble = new BlobIdPreamble(stream);
```

Comment:
You could potentially just store the BlobIdPreamble as a class member instead of extracting it. Up to you though, since that's mostly cosmetic.

Reply:
I think the problem is that `BlobId` also gets constructed in ways that do not use a `DataInputStream` (like when new ids are constructed) and to support that I would have to introduce a constructor for the preamble that takes in these fields. That would be more of a burden and an invasive change for this class I think.

Reply:
ah, I didn't think about that constructor. that makes sense